Add mqtt to LG ThinQ integration

LG-ThinQ-Integration opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add mqtt to LG ThinQ integration

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

JangwonLeeLG committed:
Add mqtt to LG ThinQ integration

JangwonLeeLG committed:
Add mqtt to LG ThinQ integration

LG-ThinQ-Integration committed:
Merge branch 'home-assistant:dev' into pr3_mqtt

LG-ThinQ-Integration committed:
Merge branch 'dev' into pr3_mqtt

farmio committed:
Update knx-frontend to 2024.9.10.221729 (#125734)

joostlek commented on homeassistant/components/lg_thinq/__init__.py:
```diff
@@ -3,22 +3,41 @@
+class ThinqData:
+    """A class that holds runtime data."""
+
+    coordinator_map: dict[str, DeviceDataUpdateCoordinator] = field(
```
```suggestion
    coordinators: dict[str, DeviceDataUpdateCoordinator] = field(
```

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lg_thinq/__init__.py:
```diff
@@ -3,22 +3,41 @@
+    coordinator_map: dict[str, DeviceDataUpdateCoordinator] = field(
+        default_factory=dict
+    )
+    mqtt_client: ThinQMQTT | None = None
```
The class is always created right? Should we make it not nullable?

LG-ThinQ-Integration replied:
What if mqtt is not working, but the other cloud functions are working well temporarily?
Do we have to stop other functions?
If we want to keep the functions to work, we should allow None for 'mqtt_client'.

joostlek commented on homeassistant/components/lg_thinq/const.py:
```diff
@@ -10,3 +11,10 @@
 CONF_CONNECT_CLIENT_ID: Final = "connect_client_id"
+
+# MQTT
+MQTT_SUBSCRIPTION_INTERVAL: Final = timedelta(days=180)
```
Why do we set it to this high value? I don't expect a majority running their HA for 180 days straight

LG-ThinQ-Integration replied:
Changed to 1.

joostlek commented on homeassistant/components/lg_thinq/mqtt.py:
```diff
@@ -0,0 +1,165 @@
+
+            # Connect to server and create certificate.
+            return await self.client.async_prepare_mqtt()
+        except Exception as exc:  # noqa: BLE001
```
Can we catch more specific exceptions? We only allow the config flow to catch bare Exceptions

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lg_thinq/mqtt.py:
```diff
@@ -0,0 +1,165 @@
+        if self.client is not None:
+            try:
+                await self.client.async_disconnect()
+            except Exception as exc:  # noqa: BLE001
```
idem

LG-ThinQ-Integration replied:
idem

joostlek commented on homeassistant/components/lg_thinq/mqtt.py:
```diff
@@ -0,0 +1,165 @@
+            ]
+            if tasks:
+                await asyncio.gather(*tasks)
+        except Exception as exc:  # noqa: BLE001
```
idem

LG-ThinQ-Integration replied:
idem

joostlek commented on homeassistant/components/lg_thinq/mqtt.py:
```diff
@@ -0,0 +1,165 @@
+            try:
+                await self.client.async_disconnect()
+            except Exception as exc:  # noqa: BLE001
+                _LOGGER.error("Failed to disconnect: %s", exc)
```
```suggestion
                _LOGGER.exception("Failed to disconnect: %s", exc)
```
This will log the stacktrace as well

joostlek commented on homeassistant/components/lg_thinq/mqtt.py:
```diff
@@ -0,0 +1,165 @@
+    async def async_connect(self) -> bool:
+        """Create a mqtt client and then try to connect."""
+        try:
+            self.client = await ThinQMQTTClient(
+                self.thinq_api, self.client_id, self.on_message_received
+            )
+            if self.client is None:
+                return False
```
Can this raise? We should only have stuff in the try block that can raise

LG-ThinQ-Integration replied:
Modified

epenet committed:
Bump sfrbox-api to 0.0.11 (#125732)

* Bump sfrbox-api to 0.0.11

* Re-enable tests

noahhusby committed:
Bump aiostreammagic to 2.2.3 (#125704)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

tl-sl committed:
Add server side events to Smlight integration (#125553)

* Register SSE client

* Add switch events for settings changes

* Mock sse settings events

* Apply suggestions from code review

Co-authored-by: Paarth Shah <mail@shahpaarth.com>

* access callbacks from mock call_args

---------

Co-authored-by: Paarth Shah <mail@shahpaarth.com>

silamon committed:
Allow to play a LinkPlay preset (#125204)

* Allow to play a linkplay preset

* Make it an entity service

* Fixes

* PR feedback

* Rename more

agoode committed:
Take a list of values for testing Threshold (#125705)

When parameterizing these tests, I forgot that hysteresis tests are
sensitive to all previous values rather than just the previous one.

This change should restore behavior to the pre-parameterization version
by replaying all value histories.

Subsequent changes will add new test cases.

mrpasztoradam committed:
Add ADS valve integration (#125619)

* feat: Add ADS valve integration

* fix: replace imports to adhere with #125665

* fix: address review feedback.

* fix: address review feedback.

catsmanac committed:
Improve test code coverage for enphase_envoy (#125582)

* Improve test code coverage for enphase_envoy

* Update tests/components/enphase_envoy/test_init.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

hunterjm committed:
Duke Energy Integration (#125489)

* Duke Energy Integration

* add recorder mock fixture to all tests

* address PR comments

* update tests

* add basic coordinator tests

* PR comments round 2

* Fix

---------

Co-authored-by: Joostlek <joostlek@outlook.com>

catsmanac committed:
Add storage settings for enphase_envoy batteries without enpower device (#125527)

* Add battery storage settings for enphase_envoy EU configuration

* Add EU Battery test fixture to enphase_envoy

* Add tests and snapshots for enphase_envoy EU battery

* refactor eu battery fixture to align with other enphase_envoy fixtures

* remove if from test and use test parameter for eu battery enphase_envoy tests

yyolk committed:
Update worldclock component config_flow labels to match pre-defined format output (#125707)

update labels for pre-defined options

update labels for pre-defined options to match strftime's formatted output

epenet committed:
Simplify imports in remote_rpi_gpio (#125745)

epenet committed:
Use HassKey in trace (#125751)

epenet committed:
Add flexibility to HassEnforceClassModule (#125739)

* Add flexibility to HassEnforceClassModule

* Adjust

epenet committed:
Simplify imports in mysensors (#125746)

akrabi committed:
Bump russound to 0.2.0 (#125743)

* Update russound library to fix BrokenPipeError

* Remove library from license expection list

rvmey committed:
Add TRIGGERcmd integration (#121268)

* Initial commit with errors

* Commitable

* Use triggercmd user id as hub name

* Validate the token

* Use switch type, no trigger yet

* Working integration

* Use triggercmd module instead of httpx

* Add tests for triggercmd integration

* Add triggercmd to requirements_test_all.txt

* Add untested triggercmd files to .coveragerc

* Implement cgarwood's PR suggestions

* Address PR feedback

* Update homeassistant/components/triggercmd/config_flow.py

Co-authored-by: Robert Resch <robert@resch.dev>

* Update homeassistant/components/triggercmd/hub.py

Co-authored-by: Robert Resch <robert@resch.dev>

* Update homeassistant/components/triggercmd/strings.json

Co-authored-by: Robert Resch <robert@resch.dev>

* Update homeassistant/components/triggercmd/hub.py

Co-authored-by: Robert Resch <robert@resch.dev>

* Get user id via triggercmd module, and better check for status 200 code

* PR feedback fixes

* Update homeassistant/components/triggercmd/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/triggercmd/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* More PR feedback fixes

* Update homeassistant/components/triggercmd/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/triggercmd/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/triggercmd/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* More PR feedback fixes

* Update tests/components/triggercmd/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Changes for PR feedback

* Changes to address PR comments

* Fix connection error when no internet

* Update homeassistant/components/triggercmd/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/triggercmd/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/triggercmd/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/triggercmd/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update tests/components/triggercmd/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Updates for PR feedback

* Update tests/components/triggercmd/test_config_flow.py

---------

Co-authored-by: Robert Resch <robert@resch.dev>
Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

epenet committed:
Use HassKey in ads (#125735)

starkillerOG committed:
Fix favorite position missing for Motion Blinds TDBU devices (#125750)

* Add favorite position for TDBU

* fix styling

epenet committed:
Move shared constant in ios (#125748)

epenet committed:
Simplify imports in pilight (#125747)

epenet committed:
Move shared constant in zabbix (#125744)

noahhusby committed:
Add seek, shuffle, and repeat controls to Cambridge Audio (#125758)

* Add advanced transport controls to Cambridge Audio

* Use TransportControl model for play/pause

jpbede committed:
Remove deprecated attributes from ping binary sensor (#125760)

joostlek committed:
Remove commented out code from weatherflow cloud (#125759)

jpbede committed:
Remove deprecated `ring.update` action (#125762)

jonnynch committed:
Bump to python-nest-sdm to 5.0.1 (#125706)

jeeftor committed:
Add Monarch Money Integration (#124014)

* Initial commit

* Second commit - with some coverage but errors abount

* Updated testing coverage

* Should be just about ready for PR

* Adding some error handling for wonky acocunts

* Adding USD hardcoded as this is all that is currently supported i believe

* updating snapshots

* updating entity descrition a little

* Addign cashflow in

* adding aggregate sensors

* tweak icons

* refactor some type stuff as well as initialize the pr comment addressing process

* remove empty fields from manifest

* Update homeassistant/components/monarchmoney/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* move stuff

* get logging out of try block

* get logging out of try block

* using Subscription ID as stored in config entry for unique id soon

* new unique id

* giving cashflow a better unique id

* Moving subscription id stuff into setup of coordinator

* Update homeassistant/components/monarchmoney/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* ruff ruff

* ruff ruff

* split ot value and balance sensors... need to go tos leep

* removed icons

* Moved summary into a data class

* efficenty increase

* Update homeassistant/components/monarchmoney/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/monarchmoney/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/monarchmoney/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/monarchmoney/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* refactor continues

* removed a comment

* forgot to add a little bit of info

* updated snapshot

* Updates to monarch money using the new typed/wrapper setup

* backing lib update

* fixing manifest

* fixing manifest

* fixing manifest

* Version 0.2.0

* fixing some types

* more type fixes

* cleanup and bump

* no check

* i think i got it all

* the last thing

* update domain name

* i dont know what is in this commit

* The Great Renaming

* Moving to dict style accounting - as per request

* updating backing deps

* Update homeassistant/components/monarch_money/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update tests/components/monarch_money/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update tests/components/monarch_money/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update tests/components/monarch_money/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/monarch_money/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* some changes

* fixing capitalizaton

* test test test

* Adding dupe test

* addressing pr stuff

* forgot snapshot

* Fix

* Fix

* Update homeassistant/components/monarch_money/sensor.py

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

epenet committed:
Simplify imports in mqtt (#125749)

epenet committed:
Move overkiz cover definitions (#125757)

epenet committed:
Move overkiz water heater definitions (#125756)

epenet committed:
Move overkiz climate definitions (#125741)

austinmroczek committed:
Add TotalConnect option to require alarm code (#122270)

* add config option

* use code_required option in alarm

* test code_required options

* only use code for disarm

* change tests to disarm with code

* remove unneeded code variable

* Update homeassistant/components/totalconnect/alarm_control_panel.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* use ServiceValidationError

* translate ServiceValidationError

* complete typing

* Update tests/components/totalconnect/test_alarm_control_panel.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* use ServiceValidationError in test

* grab usercode from correct spot

* use client code instead of unfilled location code

* Revert "remove unneeded code variable"

This reverts commit 220de0e698e5779fcd7c45bee999a60ad186ab7f.

* remove unneeded code variable

* improve usercode checking

* use freezer

* fix usercode test data

* Update homeassistant/components/totalconnect/strings.json

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* Update homeassistant/components/totalconnect/strings.json

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* update test with new message

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: G Johansson <goran.johansson@shiftit.se>

silamon committed:
Small improvements to linkplay from reviews (#125766)

Small improvements

tsvi committed:
Implement time triggers with offset for timestamp sensors (#120858)

* Implement time triggers with offset for timestamp sensors

* Fix bad change

* Add testcase for multiple conf_at with offsets

* Fix fixture rename

* Fix testcase - if no offset provided, it should be just the string of the entity id

* Get test to pass

* Simplify code

* Update the messaging and make the offset optional allowing specifying only the entity_id

* Move state tracking one level up

* Implement requesteed changes

catsmanac committed:
Reload enphase_envoy integration upon envoy firmware change detection (#124650)

* Reload enphase_envoy integration upon envoy firmware change detection.

* remove persistant notification

sdb9696 committed:
Add test snapshots to ring switch and siren platforms (#125771)

Klikini committed:
Allow attaching additional data to schedule helper blocks (#116585)

* Add a new optional "data" key when defining time ranges for the schedule component that exposes the provided data in the state attributes of the schedule entity when that time range is active

* Exclude all schedule entry custom data attributes from the recorder (with tests)

* Fix setting schedule attributes to exclude from recorder, update test to verify the attributes exist but are not recorded

* Fix test to ensure schedule data attributes are not recorded

* Use vol.Any in place of vol.Or

Co-authored-by: Erik Montnemery <erik@montnemery.com>

* Remove schedule block custom data shorthand
as requested in https://github.com/home-assistant/core/pull/116585#pullrequestreview-2280260436

* Update homeassistant/components/schedule/__init__.py

---------

Co-authored-by: Erik Montnemery <erik@montnemery.com>

joostlek committed:
Bump jaraco.abode to 6.2.0 (#125512)

* Bump jaraco.abode to 6.2.0

* Bump jaraco.abode to 6.2.0

noahhusby committed:
Add media player test to Cambridge Audio (#125780)

* Add media player tests to Cambridge Audio

* Add media player tests to Cambridge Audio

* Remove unnecessary test case

* Move state_update call out of mock

* Update tests/components/cambridge_audio/test_media_player.py

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LouisChrist committed:
Bump pyblu to 1.0.2 (#125784)

LouisChrist committed:
Remove call to removed function in bluesound integration (#125779)

* Remove async_trigger_sync_on_all

* Use cast instead of instanceof

cnico committed:
Refactoring flipr integration to prepare Hub device addition (#125262)

* Addition of hub device

* coordinator udata updated after a hub action

* Unit tests update

* Unit tests improvements

* addition of tests on select and switch platforms

* wording

* Removal of select platform for PR containing only one platform

* Remove hub to maintain only the refactoring that prepare the hub device addition

* Review corrections

* wording

* Review corrections

* Review corrections

* Review corrections

JangwonLeeLG committed:
Rename coordinator_map to coordinators

JangwonLeeLG committed:
Rename coordinator_map to coordinators

LG-ThinQ-Integration committed:
Update homeassistant/components/lg_thinq/mqtt.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration committed:
Update homeassistant/components/lg_thinq/mqtt.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

noahhusby committed:
Bump aiostreammagic to 2.2.5 (#125792)

LG-ThinQ-Integration committed:
Add binary_sensor platform to LG ThinQ integration (#125664)

* Add binary_sensor platform to LG ThinQ integration

* Update homeassistant/components/lg_thinq/binary_sensor.py

* Remove unused translation key

* Add one_touch_filter property to binary_sensor

* Add one_touch_filter to icons, strings

* Update homeassistant/components/lg_thinq/strings.json

---------

Co-authored-by: jangwon.lee <jangwon.lee@lge.com>
Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

synesthesiam committed:
Fix audio format for VoIP (#125785)

Fix audio format

epenet committed:
Move growatt_server sensor definitions (#125755)

epenet committed:
Move sunweg sensor definitions (#125754)

emontnemery committed:
Disable ESPHome assist_in_progress binary sensor (#125802)

bieniu committed:
Add support for cover tilt for Shelly 2PM Gen3 (#125717)

* Add support for tilt

* Fix config

* Add test

* Increase test coverage

gjohansson-ST committed:
Validate target temp features in Climate Entity (#125180)

* Validate target temp features in Climate Entity

* Soften

* Break long string

HarvsG committed:
Accept more than 1 state for numeric entities in Bayesian (#119281)

* test driven delevopment

* test driven development - multi numeric state

* better multi-state processing

* when state==below return true

* adds test for a bad state

* improve codecov

* value error already handled in async_numeric_state

* remove whitespace

* remove async_get

* linting

* test_driven dev for error handling

* make tests fail correctly

* ensure tests fail correctly

* prevent bad numeric entries

* ensure no overlapping ranges

* fix tests, as error caught in validation

* remove redundant er call

* remove reddundant arg

* improves code coverage

* filter for numeric states before testing overlap

* adress code review

* skip non numeric configs but continue

* wait to avoid race condition

* Better tuples name and better guard clause

* better test description

* more accurate description

* Add comments to calculations

* using typing not collections as per ruff

* Apply suggestions from code review

Co-authored-by: Erik Montnemery <erik@montnemery.com>

* follow on from suggestions

* Lazy evaluation

Co-authored-by: Erik Montnemery <erik@montnemery.com>

* update error text in tests

* fix broken tests

* move validation function call

* fixes return type of above_greater_than_below.

* improves codecov

* fixes validation

---------

Co-authored-by: Erik Montnemery <erik@montnemery.com>

sdb9696 committed:
Finish cleanup of deprecated ring update service (#125810)

mvdwetering committed:
Remove manufacturer name from Wake on LAN device_info (#123836)

Remove made up  manufacturer

epenet committed:
Use root import for media_player and media_source in tests (#125829)

epenet committed:
Improve type hints in ads (#125825)

* Improve type hints in ads

* One more

* Adjust

epenet committed:
Use root import for media_player and media_source (#125828)

* Use root import for media_player and media_source

* One more

noahhusby committed:
Add entity available attribute to Cambridge Audio (#125831)

* Bump aiostreammagic to 2.2.4

* Move callback handling to entity class

* Wrap all module exceptions in HA errors for Cambridge Audio

epenet committed:
Use root import for device_automation (#125836)

epenet committed:
Use root import for websocket_api (#125834)

jpbede committed:
Remove deprecated YAML import from rova (#125849)

emontnemery committed:
Disable Wyoming assist_in_progress binary sensor (#125806)

emontnemery committed:
Disable voip call_in_progress binary sensor (#125812)

epenet committed:
Use root import for async_redact_data in diagnostics (#125821)

gjohansson-ST committed:
Improve logging message for validation in climate (#125837)

epenet committed:
Use root import in components (#125858)

epenet committed:
Fix incorrect import in androidtv tests (#125860)

JangwonLeeLG committed:
Merge branch 'pr3_mqtt' of https://github.com/LG-ThinQ-Integration/core into pr3_mqtt

JangwonLeeLG committed:
Update for reviews

LG-ThinQ-Integration commented:
I am sorry for the inconvenience. I made a mistake when I rebase.
